File: crates/piston-nn/src/loss.rs
use piston::{DType, ScopePusher, Tensor};

pub fn nll(inp: Tensor, target: Tensor) -> anyhow::Result<Tensor> {
    let _scope_guard = ScopePusher::new("loss:nll");
    let b_sz = match &target.shape().to_vec()[..] {
        &[b_sz] => b_sz,
        dims => anyhow::bail!("the target tensor should have a single dimension ({dims:?})"),
    };
    match &inp.shape().to_vec()[..] {
        &[inp_b_sz, _] => {
            if inp_b_sz != b_sz {
                anyhow::bail!("batch size mismatch between inp ({inp_b_sz}) and target ({b_sz})")
            }
        }
        dims => anyhow::bail!("the target tensor should have two dimensions ({dims:?})"),
    }
    inp.gather(target.clone().unsqueeze(1)?, 1)?
        .affine(-1f32 / b_sz as f32, 0.)
}

pub fn nll_masked(inp: Tensor, target: Tensor) -> anyhow::Result<Tensor> {
    let _scope_guard = ScopePusher::new("loss:nll_masked");
    let b_sz = match &target.shape().to_vec()[..] {
        &[b_sz] => b_sz,
        dims => anyhow::bail!("the target tensor should have a single dimension ({dims:?})"),
    };
    match &inp.shape().to_vec()[..] {
        &[inp_b_sz, _] => {
            if inp_b_sz != b_sz {
                anyhow::bail!("batch size mismatch between inp ({inp_b_sz}) and target ({b_sz})")
            }
        }
        dims => anyhow::bail!("the target tensor should have two dimensions ({dims:?})"),
    }

    // We do a bunch of work here to allow ignoring tokens in the target.
    let ignore_index = -100;
    let mask = target.clone().ne(Tensor::full(
        target.shape(),
        ignore_index,
        target.device(),
        false,
    )?)?;

    // Note here that we seem to be able to get away with passing negative indices to gather.
    // If we were more careful about this, we'd replace the indices with 0s where the mask is 0,
    // before passing them to gather.

    let per_sample_loss = inp
        .gather(target.clone().unsqueeze(1)?, 1)?
        .affine(-1f32, 0.)?;
    let mask_unsqueezed = mask.clone().unsqueeze(1)?;
    let masked_loss = per_sample_loss
        .clone()
        .mul(mask_unsqueezed.cast(DType::F32)?)?;

    let valid_count = mask.cast(DType::F32)?.sum_all()?;

    masked_loss.div(valid_count.cast(DType::F32)?)
}

/// Computes label-smoothed cross-entropy for a flattened `[batch_size, vocab_size]` log-softmax
/// tensor `log_probs`, together with a corresponding `target` of shape `[batch_size]`.
///
/// `alpha` is the smoothing parameter in `[0, 1]`. If `alpha == 0.0`, this is just ordinary NLL.
pub fn label_smoothed_nll(log_probs: Tensor, target: Tensor, alpha: f32) -> anyhow::Result<Tensor> {
    let _scope_guard = ScopePusher::new("loss:nll_label_smoothed");
    let b_sz = match &target.shape().to_vec()[..] {
        &[b_sz] => b_sz,
        dims => {
            anyhow::bail!("label_smoothed_nll: target must be [batch_size], got shape {dims:?}")
        }
    };
    let shape_lp = log_probs.shape().to_vec();
    if shape_lp.len() != 2 {
        anyhow::bail!(
            "label_smoothed_nll: log_probs must be rank-2 [batch_size, vocab_size], got {shape_lp:?}"
        );
    }
    let (inp_b_sz, vocab_size) = (shape_lp[0], shape_lp[1]);
    if inp_b_sz != b_sz {
        anyhow::bail!(
            "label_smoothed_nll: batch size mismatch between log_probs ({inp_b_sz}) and target ({b_sz})"
        );
    }

    // Check for ignored tokens (often `-100` in NLP).
    let ignore_index = -100;
    let mask = target
        .clone()
        .ne(Tensor::full(
            target.shape(),
            ignore_index,
            target.device(),
            false,
        )?)?
        .cast(piston::DType::F32)?;

    // Gather the negative log-prob for the correct class:
    // nll_loss[i] = -log_probs[i, target[i]]  (for each token i).
    let nll_gathered = log_probs
        .clone()
        .gather(target.clone().unsqueeze(1)?, 1)? // shape [batch_size, 1]
        .affine(-1.0, 0.0)?; // multiply by -1

    // Mask out ignored tokens (multiply by 0 where masked=0).
    let nll_masked = nll_gathered.mul(mask.clone().unsqueeze(1)?)?;

    // We'll also average over only the valid tokens:
    let valid_count = mask.clone().sum_all()?; // shape []

    // (1) Ordinary cross-entropy term (averaged).
    let nll_loss = nll_masked.div(valid_count.clone())?;

    // (2) Uniform penalty term, also masked.
    //
    // For label smoothing, we pretend a small fraction α of the time
    // we want the “average” log-prob over all classes, not just the correct one.
    // i.e. uniform_loss = - average over (vocab_size) of log_probs, restricted to masked positions.
    let all_probs_masked = log_probs.mul(mask.unsqueeze(1)?)?;
    let sum_log_probs = all_probs_masked.sum(1)?;
    // Now shape [batch_size], each entry is sum_{v in vocab} log_probs[i, v].
    // Negative average per token:
    let neg_avg_log_prob = sum_log_probs.affine(-1.0 / vocab_size as f32, 0.0)?;
    let uniform_loss = neg_avg_log_prob.sum_all()?.div(valid_count)?;

    // Combine with alpha
    // final = (1 - alpha) * nll + alpha * uniform_term
    let final_loss = nll_loss
        .affine(1.0 - alpha, 0.0)?
        .add(uniform_loss.affine(alpha, 0.0)?)?;

    Ok(final_loss)
}

pub fn log_softmax(xs: Tensor, d: usize) -> anyhow::Result<Tensor> {
    let _scope_guard = ScopePusher::new("loss:log_softmax");
    let max = xs.clone().max_keepdim(d)?;
    let diff = xs.clone().sub(max)?;
    let sum_exp = diff.clone().exp()?.sum_keepdim(d)?;
    let log_sm = diff.sub(sum_exp.log()?)?;
    Ok(log_sm)
}

pub fn cross_entropy(inp: Tensor, target: Tensor) -> anyhow::Result<Tensor> {
    let _scope_guard = ScopePusher::new("loss:cross_entropy");
    if inp.rank() != 2 {
        anyhow::bail!("cross_entropy expects an input tensor of rank 2")
    }
    let inp = log_softmax(inp, 1)?;
    nll(inp, target)?.sum_all()
}

#[cfg(all(test, feature = "pyo3"))]
mod tests {
    use super::*;
    use anyhow::Result;
    use piston::{DType, Device, DeviceRequest, Parameter};
    use test_strategy::{proptest, Arbitrary};

    thread_local! {
        static GPU_DEVICE: Device = Device::request_device(DeviceRequest::GPU).unwrap();
    }

    fn ground_truth_cross_entropy(input: &Tensor, target: &Tensor) -> Result<(Tensor, Tensor)> {
        let grad_prg = r#"
import torch
import torch.nn.functional as F

def cross_entropy_with_grad(input, target):
    input_tensor = torch.tensor(torch.from_numpy(input), requires_grad=True)
    target_tensor = torch.tensor(torch.from_numpy(target), dtype=torch.long)
    loss = F.cross_entropy(input_tensor, target_tensor, reduction='mean')
    loss.backward()
    return input_tensor.grad.numpy()
"#;
        let grad =
            piston::test_util::run_py_prg(grad_prg.to_string(), &[input, target], &[], DType::F32)?;

        let loss_prg = r#"
import torch
import torch.nn.functional as F

def cross_entropy(input, target):
    input_tensor = torch.tensor(torch.from_numpy(input))
    target_tensor = torch.tensor(torch.from_numpy(target), dtype=torch.long)
    return F.cross_entropy(input_tensor, target_tensor, reduction='mean').numpy()
"#;
        let loss =
            piston::test_util::run_py_prg(loss_prg.to_string(), &[input, target], &[], DType::F32)?;

        Ok((loss, grad))
    }

    #[derive(Arbitrary, Debug)]
    struct CrossEntropyProblem {
        #[strategy(1..=32usize)]
        batch_size: usize,
        #[strategy(2..=10usize)]
        num_classes: usize,
    }

    fn run_cross_entropy_trial(problem: CrossEntropyProblem) -> Result<()> {
        let device = GPU_DEVICE.with(|d| d.clone());
        let CrossEntropyProblem {
            batch_size,
            num_classes,
        } = problem;

        // Generate random input and target tensors
        let input = Tensor::randn::<f32, _>(0., 1., (batch_size, num_classes), Device::CPU)?;
        let target = Tensor::randint(0, num_classes as i32, batch_size, Device::CPU)?;

        // Compute ground truth
        let (ground_loss, ground_grad) = ground_truth_cross_entropy(&input, &target)?;

        // Compute our implementation
        let input_gpu = input.to(&device)?;
        let target_gpu = target.to(&device)?;
        let input_param = Parameter::from_tensor(&input_gpu)?;
        let our_loss = cross_entropy(input_param.as_tensor().clone(), target_gpu)?;

        // Compute gradients
        our_loss.backward()?;
        device.try_gpu()?.mark_step()?;
        let our_grad = input_param.as_tensor().grad().unwrap().clone();

        // Compare results
        let our_loss_cpu = our_loss.to(&Device::CPU)?;
        let our_grad_cpu = our_grad.to(&Device::CPU)?;
        let ground_grad = ground_grad.to(&Device::CPU)?;
        let ground_loss = ground_loss.to(&Device::CPU)?;

        println!("Input shape: {:?}", input.shape());
        println!("Target shape: {:?}", target.shape());
        println!("Our loss: {:?}", our_loss_cpu.to_vec::<f32>());
        println!("Ground truth loss: {:?}", ground_loss.to_vec::<f32>());
        println!("Our grad: {:?}", our_grad_cpu.to_vec::<f32>());
        println!("Ground truth grad: {:?}", ground_grad.to_vec::<f32>());
        println!("Our grad shape: {:?}", our_grad_cpu.shape());
        println!("Ground truth grad shape: {:?}", ground_grad.shape());

        ground_loss.all_close(&our_loss_cpu, 1e-5, 1e-5)?;
        ground_grad.all_close(&our_grad_cpu, 1e-5, 1e-5)?;

        Ok(())
    }

    #[proptest(cases = 10)]
    fn test_cross_entropy(prob: CrossEntropyProblem) {
        let CrossEntropyProblem {
            batch_size,
            num_classes,
        } = prob;
        println!(
            "Testing with batch_size = {}, num_classes = {}",
            batch_size, num_classes
        );
        run_cross_entropy_trial(prob).unwrap();
    }
}

File: packages/web/src/nn/loss.ts
import { float32, full, int32 } from "@/globals";
import { LogSoftmax } from "@/nn/activation";
import { Module } from "@/nn/module";
import { Tensor } from "@/tensor";

/**
 * Configuration options for CrossEntropyLoss
 * @interface CrossEntropyLossConfig
 */
export interface CrossEntropyLossConfig {
  /**
   * Index to ignore in the target tensor
   * @default -100
   */
  ignoreIndex?: number;

  /**
   * Specifies the reduction to apply to the output
   * @default 'mean'
   */
  reduction?: "none" | "sum" | "mean";

  /**
   * Smoothing factor for label smoothing regularization
   * @default 0.0
   */
  labelSmoothing?: number;
}

export class CrossEntropyLoss extends Module {
  private ignoreIndex: number;
  private reduction: "none" | "sum" | "mean";
  private labelSmoothing: number;
  private logSoftmax: LogSoftmax;

  constructor(config: CrossEntropyLossConfig = {}) {
    super();
    this.ignoreIndex = config.ignoreIndex ?? -100;
    this.reduction = config.reduction ?? "mean";
    this.labelSmoothing = config.labelSmoothing ?? 0.0;
    this.logSoftmax = new LogSoftmax(1);
  }

  forward(input: Tensor, target: Tensor): Tensor {
    // Validate shapes
    const inputShape = input.size();
    const targetShape = target.size();

    if (inputShape.length !== 2) {
      throw new Error(
        `CrossEntropyLoss: input must be rank-2 [batch_size, vocab_size], got ${inputShape}`,
      );
    }

    if (targetShape.length !== 1) {
      throw new Error(`CrossEntropyLoss: target must be [batch_size], got ${targetShape}`);
    }

    const [batchSize, vocabSize] = inputShape;
    const targetBatchSize = targetShape[0];

    if (batchSize !== targetBatchSize) {
      throw new Error(
        `CrossEntropyLoss: batch size mismatch between input (${batchSize}) and target (${targetBatchSize})`,
      );
    }

    // Apply log softmax to the input
    const logProbs = this.logSoftmax.forward(input);

    // Create mask for ignored indices
    const mask = target
      .ne(
        full(targetShape, this.ignoreIndex, {
          device: target.device,
          dtype: int32,
        }),
      )
      .cast(float32);

    // Gather the negative log-prob for the correct classes
    const nllGathered = logProbs.gather(1, target.unsqueeze(1)).mul(-1);

    // Mask out ignored tokens
    const nllMasked = nllGathered.mul(mask.unsqueeze(1));

    // At this point, nllMasked has shape [batchSize, 1]. We squeeze it to get a per-sample loss.
    const nllLossPerSample = nllMasked.squeeze({ dim: 1 }); // shape: [batchSize]

    // Compute final per-sample loss (without global normalization)
    let finalLossPerSample = nllLossPerSample;

    // If label smoothing is applied, compute the uniform loss per sample and combine.
    if (this.labelSmoothing > 0) {
      // Calculate uniform loss: average log probability over all classes for each sample.
      const allProbsMasked = logProbs.mul(mask.unsqueeze(1));
      const sumLogProbs = allProbsMasked.sum({ dim: 1 }); // shape: [batchSize]
      const uniformLossPerSample = sumLogProbs.mul(-1 / vocabSize);

      // Combine with the nll loss using the smoothing factor.
      finalLossPerSample = nllLossPerSample
        .mul(1 - this.labelSmoothing)
        .add(uniformLossPerSample.mul(this.labelSmoothing));
    }

    // Apply reduction according to the configuration.
    if (this.reduction === "none") {
      // Return the per-sample losses directly.
      return finalLossPerSample;
    } else if (this.reduction === "sum") {
      // Sum up all per-sample losses.
      return finalLossPerSample.sum();
    } else {
      // For "mean", average the sum of losses by the count of valid tokens.
      const validCount = mask.sum();
      return finalLossPerSample.sum().div(validCount);
    }
  }
}
